vamb/h_loss.py

```diff
@@ -0,0 +1,820 @@
+"""Hierarchical loss for the labels suggested in https://arxiv.org/abs/2210.10929"""
+
+
+__cmd_doc__ = """Hierarchical loss for the labels"""
+
+import torch as _torch
+from torch import nn as _nn
+from torch.utils.data.dataset import TensorDataset as _TensorDataset
+import torch.nn.functional as F
+from typing import Optional, IO, Union
+from pathlib import Path
+from torch.utils.data.dataset import TensorDataset as _TensorDataset
+from torch.utils.data import DataLoader as _DataLoader
+from torch.optim import Adam as _Adam
+from torch import Tensor
+_torch.manual_seed(0)
+from collections import namedtuple
+
+import networkx as nx
+
+import numpy as _np
+from functools import partial
+
+from math import log as _log
+
+import vamb.semisupervised_encode as _semisupervised_encode
+import vamb.encode as _encode
+import vamb.hlosses_fast as _hlosses_fast
+import vamb.hier as _hier
+import vamb.infer as _infer
+
+if _torch.__version__ < '0.4':
+    raise ImportError('PyTorch version must be 0.4 or newer')
+
+
+def get_level(x):
+    if x.startswith('d_'):
+        return 1
+    if x.startswith('p_'):
+        return 2
+    if x.startswith('c_'):
+        return 3
+    if x.startswith('o_'):
+        return 4
+    if x.startswith('f_'):
+        return 5
+    if x.startswith('g_'):
+        return 6
+    if x.startswith('s_'):
+        return 7
+
+
+def make_graph(taxs):
```

Comment:
What's the input and output types? You can make a type alias in a global variable if it's too complicated.

Reply:
I know typing sucks when you work with tensors.